### PR TITLE
[llvm-spirv] Cherry pick Khronos changes to expand collection of entry point interfaces (PR #1334)

### PR DESCRIPTION
**Pull request by asudarsa**
This PR pulls in the following PR from upstream Khronos SPIRV-LLVM-Translator repo:
https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1334

`
This is a patch to expand the collection of entry point interfaces.
In SPIR-V 1.4 and later OpEntryPoint must list all global variables in the
interface.
`
In addition, a couple of minor changes have been added to sync with latest code.
This patch addresses https://github.com/intel/llvm/issues/9958

Updated the following tests to sync with upstream as well:
        llvm-spirv/test/extensions/INTEL/SPV_INTEL_inline_assembly/inline_asm_clobbers.cl
        llvm-spirv/test/extensions/INTEL/SPV_INTEL_inline_assembly/inline_asm_constraints.cl

Thanks

### CODE REVIEW

File: llvm-spirv/lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -874,9 +874,7 @@ SPIRVFunction *LLVMToSPIRVBase::transFunctionDecl(Function *F) {
 
     BM->setName(BF, F->getName().str());
   }
-  if (isKernel(F))
-    BM->addEntryPoint(ExecutionModelKernel, BF->getId());
-  else if (F->getLinkage() != GlobalValue::InternalLinkage)
+  if (!isKernel(F) && F->getLinkage() != GlobalValue::InternalLinkage)
```
Comment (asudarsa) [inline llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:877]: We no longer add entry point for function decls.

File: llvm-spirv/lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -4915,9 +4916,14 @@ void LLVMToSPIRVBase::collectInputOutputVariables(SPIRVFunction *SF,
     }
 
     if (isAnyFunctionReachableFromFunction(F, Funcs)) {
-      SF->addVariable(ValueMap[&GV]);
+      SPIRVWord ModuleVersion = static_cast<SPIRVWord>(BM->getSPIRVVersion());
+      if (AS != SPIRAS_Input && AS != SPIRAS_Output &&
+          ModuleVersion < static_cast<SPIRVWord>(VersionNumber::SPIRV_1_4))
+        BM->setMinSPIRVVersion(VersionNumber::SPIRV_1_4);
```
Comment (asudarsa) [inline llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:4922]: We now pass Version number directly without converting to SPIRVWord

File: llvm-spirv/lib/SPIRV/SPIRVWriter.h [hunk 3]
```diff
@@ -248,6 +248,8 @@ class LLVMToSPIRVBase : protected BuiltinCallHelper {
       const Function *FS,
       const std::unordered_set<const Function *> Funcs) const;
   void collectInputOutputVariables(SPIRVFunction *SF, Function *F);
+  std::vector<SPIRVId> collectEntryPointInterfaces(SPIRVFunction *BF,
```
Comment (asudarsa) [inline llvm-spirv/lib/SPIRV/SPIRVWriter.h:251]: This function declaration was missing in the original patch.

**Comment by AlexeySachkov** [comment]
Just curious, why can't we get this patch with a regular pulldown?

**Review by sarnex — APPROVED** [review]
lgtm from a high level, i did not review the change in depth

**Comment by asudarsa** [comment]
Hi @MrSidims

**Comment by asudarsa** [comment]
Sorry. I pushed the wrong button which caused close/reopen.

**Comment by asudarsa** [comment]
Hi @AlexeySachkov 

I think this is ready to land. Can you please help with merging when convenient?

Thanks

**Pull request merged**